### ISSUE TITLE
Client implementation for HTTP Decorators

### ISSUE THREAD

**Issue by derekm** [nestjs/nest#13898]
### Is there an existing issue that is already proposing this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe it

Writing declarative server-side HTTP calls is awesome, but then turning around and writing manual client-side HTTP calls sucks.

### Describe the solution you'd like

I would like to re-use declarative server-side interfaces for client-side web clients.

### Teachability, documentation, adoption, migration strategy

If the server-side decorators have enough information to receive HTTP calls, then they also have enough information to send HTTP calls.

Once you have an awesome decorated server-side interface, imagine re-using that interface to generate a client, as in:

```js
const endpointClient = new ClientFromInterface<IServiceEndpoint>();
const resourceObject = endpointClient.resource().subresource(withPathParam).get()
```

### What is the motivation / use case for changing the behavior?

.NET has Refit: https://github.com/reactiveui/refit

Android has Retrofit: https://github.com/square/retrofit

Java has MicroProfile Rest Client: https://github.com/eclipse/microprofile-rest-client
... and Jersey WebResourceFactory: https://javadoc.io/doc/org.glassfish.jersey.ext/jersey-proxy-client/latest/org/glassfish/jersey/client/proxy/package-summary.html

JavaScript has ???

**Comment by micalevisk** [comment]
I don't think that Nestjs (I mean, under `@nestjs/*` packages) is aiming so solve client DX issues (See #10046). But I believe that there isn't much to change in order to help lib authors to write libs to improve client integration with nestjs apps

To be honest I didn't understand your suggestion. Few free to create a post on our [Discord](http://discord.gg/nestjs) community so we can discuss it further.

**Comment by derekm** [comment]
> I don't think that Nestjs (I mean, under `@nestjs/*` packages) is aiming to address client DX issues (See #10046). But I believe that there isn't much to change in order to help lib authors to write libs to improve client integration with nestjs apps
> 
> To be honest I didn't understand your suggestion. Few free to create a post on our [Discord](http://discord.gg/nestjs) community so we can discuss it further.

Instead of putting HTTP Decorators on classes, imagine putting them on interfaces. Now service-side classes can use them to inplement endpoints, and client-side classes can use them to talk to the same endpoints.

When done correctly, this guarantees that your native JS decorator-based clients can perfectly talk to your native JS decorator-based endpoints. Look at the Java libraries, and realize that both server-side and client-side use JAX-RS annotations. Same can be done there: annotate interfaces instead of class, and have both server and client use the same interfaces.

**Comment by jmcdo29** [comment]
Typescript doesn't allow for decorators on interfaces. It's a language limitation due to the interfaces being destroyed at compile time

**Comment by micalevisk** [comment]
Got you but this is not how typescript works. We couldn't use types, in your example

Unless we have a compiler that would create things for us, of course.

**Comment by micalevisk** [comment]
I believe that you can think of several work around to that but due to the dynamic nature of nestjs apps the best approach I can think of (with REST APIs in mind) is this: generate a OpenAPI spec from your running server then use it in your client. And you can use any _openapi to typescript_ tooling for that such as https://github.com/ferdikoomen/openapi-typescript-codegen

**Comment by derekm** [comment]
@micalevisk -- That's definitely one way to do it, but having experienced both ways, re-used interfaces will always be the best approach for native-clients. NestJS HTTP Decorators plus Zod would be the best client framework in JavaScript.

**Comment by kamilmysliwiec** [comment]
Thanks for your suggestion!

There are no plans to implement it in the foreseeable future as the effort required to implement it is not justified by the expected impact, and given how frequently TypeScript AST changes, it would be hard to maintain (proven by some of the existing CLI plugins we have for swagger/graphql).

If you think your request could live outside Nest's scope, we'd encourage you to collaborate with the [community](https://discord.gg/G7Qnnhy) on publishing it as an open source package.
